src/common/type/pack.c

```diff
@@ -0,0 +1,1571 @@
+        if (packTypeData[this->tagNextType].valueMultiBit)
+        {
+            // If the value is stored following the tag (value > 1 bit)
+            if (tag & 0x8)
+            {
+                // Read low order bits of the field ID delta
+                this->tagNextId = tag & 0x3;
+
+                // Read high order bits of the field ID delta when specified
+                if (tag & 0x4)
+                    this->tagNextId |= (unsigned int)pckReadUInt64Internal(this) << 2;
+
+                // Read value
+                this->tagNextValue = pckReadUInt64Internal(this);
+            }
+            // Else the value is stored in the tag (value == 1 bit)
+            else
+            {
+                // Read low order bit of the field ID delta
+                this->tagNextId = tag & 0x1;
+
+                // Read high order bits of the field ID delta when specified
+                if (tag & 0x2)
+                    this->tagNextId |= (unsigned int)pckReadUInt64Internal(this) << 1;
+
+                // Read value
+                this->tagNextValue = (tag >> 2) & 0x3;
+            }
+        }
+        // Else the value is a single bit (e.g. boolean)
+        else if (packTypeData[this->tagNextType].valueSingleBit)
+        {
+            // Read low order bits of the field ID delta
+            this->tagNextId = tag & 0x3;
+
+            // Read high order bits of the field ID delta when specified
+            if (tag & 0x4)
+                this->tagNextId |= (unsigned int)pckReadUInt64Internal(this) << 2;
+
+            // Read value
+            this->tagNextValue = (tag >> 3) & 0x1;
+        }
+        // Else the value is multiple tags (e.g. container)
+        else
+        {
+            // Read low order bits of the field ID delta
+            this->tagNextId = tag & 0x7;
+
+            // Read high order bits of the field ID delta when specified
+            if (tag & 0x8)
+                this->tagNextId |= (unsigned int)pckReadUInt64Internal(this) << 3;
+
+            // Value length is variable so is stored after the tag
+            this->tagNextValue = 0;
+        }
+
+        // Increment the next tag id
+        this->tagNextId += this->tagStackTop->idLast + 1;
+
+        // Tag was found
+        result = true;
+    }
+
+    FUNCTION_TEST_RETURN(result);
+}
+
+/***********************************************************************************************************************************
+Read field tag
+
+Some tags and data may be skipped based on the value of the id parameter.
+***********************************************************************************************************************************/
+static uint64_t
+pckReadTag(PackRead *this, unsigned int *id, PackType type, bool peek)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(PACK_READ, this);
+        FUNCTION_TEST_PARAM_P(UINT, id);
+        FUNCTION_TEST_PARAM(ENUM, type);
+        FUNCTION_TEST_PARAM(BOOL, peek);
```

Comment:
Can you give me an idea as to what the purpose of peek really is? It's not very clear except that it can be true or false when the type is unknown but otherwise must be false.

Reply:
Peek allows you to look at the next tag without updating the internal state of the pack. So, for instance, you can look for gaps that indicate NULLs without advancing the field id. If the field id advances, then a function like pckReadNull() would fail because the current field id in the pack would be greater than the field id requested, because field ids for NULLs don't exist. It is also used to find the end of containers.